Support all_reduce a list of same-device tensors #21640

jfc4050 opened the pull request:
addresses #21640 for CPU tensors and the Gloo backend.

Questions:
- ~~currently takes `AllreduceOptions`, since all of the options are the same. Would it be better to make a new `AllreduceCoalescedOptions` class?~~
- ~~I decided to inherit from `ProcessGroupGloo::AsyncWork` instead of `AsyncAllreduceWork` to shorten the inheritance chain a bit and for consistency with existing classes. However, this means that the two `getFunction` methods are copy-pasted. Would inheriting from `AsyncAllreduceWork` be preferable?~~
- ~~should the work class be named `AsyncCoalescedAllreduceWork` or `AsyncAllreduceCoalescedWork`?~~

thank you!

jfc4050 committed:
initial commit - added allreduce_coalesced tests

    - allreduce_coalesced test cases
    - test checks
    - basic checks
    - stress tests

jfc4050 committed:
added allreduce_coalesced

jfc4050 committed:
added tests for distributed.all_reduce_coalesced

    - {"sum", "product", "min", "max"} x {"", "group", "full_group"}

jfc4050 committed:
added distributed.all_reduce_coalesced

    - currently only implemented for Gloo backend

zhaojuanmao commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -580,6 +580,65 @@ class AsyncAllreduceWork : public ProcessGroupGloo::AsyncWork {
   }
 };
 
+class AsyncAllreduceCoalescedWork: public ProcessGroupGloo::AsyncWork {
```
feel like inherit AsyncAllreduceWork will be more cleaner

jfc4050 replied:
yeah that's actually the way I implemented it at first, but wasn't sure what the preference would be. I think this way makes more sense after thinking about it some more. Changed.

zhaojuanmao commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -580,6 +580,65 @@ class AsyncAllreduceWork : public ProcessGroupGloo::AsyncWork {
+    return fn;
+  }
+
+  void allreduce(at::Tensor& tensor) {
```
could it still be vector<at::Tensor> and input size will be 1? in this way, you do not need to overwrite the allreduce() function but just inherit from AsyncAllReduceWork

jfc4050 replied:
oh good idea, changed.

zhaojuanmao commented on torch/distributed/distributed_c10d.py:
```diff
@@ -907,6 +907,49 @@ def all_reduce(tensor,
+    in all processes.
+
+    Arguments:
+        tensor (List[Tensor]): Input and output of the collective. The function
```
nit: tensors (List[Tensor])

jfc4050 replied:
oops, fixed.

zhaojuanmao reviewed (commented):
nice work! are you planning to add allreduce_coalesced cuda version as well?

jfc4050 committed:
fixed all_reduce_coalesced docstring: s/tensor/tensors/

jfc4050 committed:
AsyncAllreduceCoalescedWork inherits from AsyncAllreduceWork

    - removed getFunction copy-pastes
    - passing single-tensor vector to inherited allreduce method

jfc4050 commented:
> nice work! are you planning to add allreduce_coalesced cuda version as well?

thank you! Not sure yet. I would like to but it might be difficult for me to test since I only have 1 GPU at home. I'm thinking about just launching an EC2 instance for testing

mrshenli reviewed (commented):
Great work!! Thanks for contributing! 

Most of my comments are nits, except the one regarding enforcing tensor sizes. Please let me know if I misread your implementation.

jfc4050 committed:
simple_coalesced_reduce_tests: s/1.0/1/

    - for visual consistency

jfc4050 committed:
added an early check for tensor layout

    - catches non-strided layouts before call to nextTag
    - added tests for early checks
    - using c10::* instead at::* for devices and layouts

jfc4050 committed:
added warning doc to all_reduce_coalesced

    - warns of lack of cross-node shape checking.

jfc4050 committed:
added AllreduceCoalescedOptions

    - also fixed py::arg misnaming for allreduce_coalesced s/op/opts/

mrshenli reviewed (approved):
Thanks a lot for implementing coalesced allreduce!

Test failures are irrelevant, but let's wait a bit to see if it will get fixed on master soon.

Commit:
Merge remote-tracking branch 'origin/master' into HEAD

jfc4050 committed:
fixed missing AllreduceCoalescedOptions import

Commit:
Merge remote-tracking branch 'origin/master' into HEAD

xush6528 commented:
@mrshenli @jfc4050 

Hi, I am seeing this test case failing. 

https://circleci.com/gh/pytorch/pytorch/2697607?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

Can you take a look?

pietern commented:
@xush6528  Can you create a new issue for failing/flaky tests? This PR has already been merged a while ago, so it's unlikely we can still revert. In a new issue you can also report the base commit you used when testing. This code is being changed frequently, so it's important to know which exact commit you were on for these failing tests (or link to the PR of course).

gchanan commented on test/test_c10d.py:
```diff
@@ -77,6 +77,37 @@ def simple_reduce_tests(rank, world_size):
+    return [
+        (
+            c10d.ReduceOp.SUM,
+            [torch.Tensor([rank + 1]), torch.Tensor([(rank + 1) ** 2])],
```
just a note: new code shouldn't use `torch.Tensor`, it should use `torch.tensor`.

mrshenli replied:
is `torch.Tensor` deprecated for good, or what's the difference between `torch.Tensor` and `torch.tensor` now?

gchanan commented on torch/lib/c10d/ProcessGroup.hpp:
```diff
@@ -111,6 +111,10 @@ class ProcessGroup {
       const AllreduceOptions& opts = AllreduceOptions()) = 0;
 
+  virtual std::shared_ptr<ProcessGroup::Work> allreduce_coalesced(
+      std::vector<at::Tensor>& tensors,
```
nit: I believe this vector can be non-const.

jfc4050 replied:
sorry I don't fully understand, would you mind elaborating?

did you mean that it could be made `const`? It looks like that could be possible since `copy_` is a `const` method
